### PR TITLE
Media: Add "This Post" filter to the editor media modal.

### LINKED ISSUE #2136: Media: Include "Uploaded to this Post" filter in Media Library
Core has some UX for limiting Media Library results to attachments for a specific post:

image

It would be nice if the Calypso media library had this as well. As it stands, on older posts, it's very difficult to find the images uploaded to the specific post.

#calypso-gaps

### PR DESCRIPTION
**Pull request by iseulde**
Adds the ability to filter media in the modal to only show images that are attached to the current post. This ability is present in the core media picker & this fills that gap in Calypso.

fixes #2136

Testing instructions:

* The [media library page](http://calypso.localhost:3000/media) should not have this filter.
* A [new post](http://calypso.localhost:3000/post)'s media modal will not have this filter until the post is auto-saved.
* An existing post's media modal without any media uploaded to it should show "There are no media uploaded to this post." when the filter is enabled.
* An existing post's media modal with media uploaded to it should show these media when the filter is enabled.

Cc @folletto.

### CODE REVIEW

**Comment by folletto** [comment]
Tested in Chrome, design wise: 👍 

Visual reference:

screen shot 2017-01-07 at 20 45 25

---

Two discussion points:

> An existing post's media modal without any media uploaded to it should show "There are no media uploaded to this post." when the filter is enabled.

I'm not sure about this. It feels slightly frustrating to me to see "This Post" active and then... "No media" message. Can we just hide the tab if there's no media? Is this a technical limitation?

> An existing post's media modal with media uploaded to it should show these media when the filter is enabled.

This is ok, and it made me think if we should select it by default when available. What do you think?

**Comment by iseulde** [comment]
> Can we just hide the tab if there's no media? Is this a technical limitation?

That should be possible. Will try it.

> This is ok, and it made me think if we should select it by default when available. What do you think?

Sure, I think that's a good idea! :)

**Comment by iseulde** [comment]
> Can we just hide the tab if there's no media? Is this a technical limitation?

Actually this would require us to fetch the media before the user navigates to the tab, while currently the media are fetched afterwards.

**Comment by iseulde** [comment]
Similarly I wonder if we should hide the other filters if no content is available there. This is what core does with the drop down (and the count is showed as well). The type of filter is slightly different though.

**Comment by iseulde** [comment]
Regarding selecting it by default, do you think it should still be second in row, or should it then be first?

**Comment by iseulde** [comment]
Also, if we select "This Post" by default, it will be a weird experience when "All" is selected when there are no media for that post. What should happen when the user then uploads media? Switching to "This Post" will be disorienting.

Maybe some users will want to work in the "This Post" space, even if it's empty.

**Comment by folletto** [comment]
> Actually this would require us to fetch the media before the user navigates to the tab, while currently the media are fetched afterwards.

Ok, nevermind then. We can refine that later.

> Regarding selecting it by default, do you think it should still be second in row, or should it then be first?

Should stay as second, keeping the hierarchy with "All" as the main container to the edge.

>  What should happen when the user then uploads media? Switching to "This Post" will be disorienting.

I wouldn't switch it. It should be a default selection when the Media modal opens, but shouldn't do anything automated once it's open. If it's on All, it'll stay on All, and so on.

> Maybe some users will want to work mainly in the "This Post" space, even if it's empty.

Yep it's possible. Ok let's make it default, and show it empty if it's empty. The only potential drawback from that is people that might feel "all my content has disappeared", so maybe we should tweak the message there?

Like:

screen shot 2017-01-07 at 21 29 29

Makes sense?

**Comment by iseulde** [comment]
Sure, adjusting the message makes sense. Maybe we should give a link or a button to access all media too? The message says there are two options, but afterwards only one is given.

We can certainly also remove the filter when there are no media for that filter, I'm just not sure if it's a good idea to combine it then with selecting it by default.

**Comment by folletto** [comment]
> Maybe we should give a link or a button to access all media too? The message says there are two options, but afterwards only one is given.

I'd be more inclined toward not having any button there, as the button is already right there. :)

> We can certainly also remove the filter when there are no media for that filter, I'm just not sure if it's a good idea to combine it then with selecting it by default.

Yup.

**Comment by iseulde** [comment]
> I'd be more inclined toward not having any button there, as the button is already right there. :)

That's true. Either there should be nothing or both, in my opinion. :)

File: client/my-sites/media-library/filter-bar.jsx [hunk 1]
```diff
@@ -40,6 +43,9 @@ module.exports = React.createClass( {
 		var placeholderText;
 
 		switch ( this.props.filter ) {
+			case 'this-post':
+				placeholderText = this.translate( 'Search media uploaded to this post…', { textOnly: true } );
```
Comment (yoavf) [inline client/my-sites/media-library/filter-bar.jsx]: textOnly is no longer necessary and can be safely removed.

**Comment by drw158** [comment]
Wow, so great to see someone already working on this!

We need to make sure this doesn't interfere with the new Media in the sidebar.

We should keep the other filters in there, even if there is no content. This allows users to know they are able to upload those types of media.

While this is a good first step, and could be shipped, I think this could be a better solution:

screen shot 2017-01-23 at 3 35 38 pm

The Me/Everyone segmented control being compared to the This Post/All filter

It could be `This post` `All posts`

**Comment by folletto** [comment]
> We should keep the other filters in there, even if there is no content. This allows users to know they are able to upload those types of media.

My assumption here would be that a user expects to upload _any_ file, it's our limitation that we filter some file types and we tier them by plan (https://en.support.wordpress.com/accepted-filetypes/). Also it's a current Calypso pattern to hide items on the filter bar if we are able to detect they are empty.
So.. maybe? But I'm not convinced.

> We need to make sure this doesn't interfere with the new Media in the sidebar.

Excellent point. /cc @iamtakashi 

> The Me/Everyone segmented control being compared to the This Post/All filter

:facepalm: How haven't I see that? :|
That's far better. Thanks!

**Comment by retrofox** [comment]
Issue: #2136

**Comment by gwwar** [comment]
I think @drw158 idea for a toggle filter vs a new tab makes a lot of sense. Especially when uploading new media items.

A few things I noticed
- The tab still says "This post" when editing a Page. Are there any custom post types we should consider as well for the control copy?
- I published a post with media inserted, but the filter didn't appear to work
- Should we show media items inserted into a draft, or only the published items?

**Comment by retrofox** [comment]
> I published a post with media inserted, but the filter didn't appear to work

I think the rule is that the images must be uploaded when the page/post is being edited. It means that the media object will have defined a parent_id into its properties.
If I'm not wrong, If the post has added an image into its content which wasn't uploaded from the post/page editor ... this image won't bellow to the post object. Crazy but true.

**Comment by gwwar** [comment]
> this image won't bellow to the post object. Crazy but true.

Oh that sucks. I suppose we could do something hacky like parse the post content for the ids

**Comment by iamtakashi** [comment]
> The tab still says "This post" when editing a Page. Are there any custom post types we should consider as well for the control copy?

This is tricky... I don't think we want to allow CPT name such as "Testimonial" or "Portfolio" to be there because that will be lengthy. If we can come up with something name agnostic, it'd be great but I'm not sure what can be... "Entry" is a bit meh, isn't it. hmmm...

**Comment by drw158** [comment]
`This item` `All items` is the only one I can think of. If it is a Post or Page though, we should definitely use those specific words instead.

**Comment by folletto** [comment]
Note that in other languages that string might get fairly long already, so length has to be taken into consideration ("This article" could get to "Questo articolo" and "All times" as "Tutti gli oggetti" / "Tutti i documenti").

**Comment by folletto** [comment]
What's missing to merge this in? I think we should merge the functionality and then refine the design later. I don't think it's a reason to hold this up – if the latest discussion above was a blocker, at least. :)

**Comment by drw158** [comment]
Is this ready to merge? Currently this Calypso gap is assigned to Triton :)

**Comment by jblz** [comment]
I chatted with @iseulde & am going to take this on.

The rebase was fairly straightforward, continuing to test & review.

File: client/my-sites/media-library/filter-bar.jsx [hunk 2]
```diff
@@ -63,6 +66,8 @@ export class MediaLibraryFilterBar extends Component {
 		const { translate } = this.props;
 
 		switch ( filter ) {
+			case 'this-post':
+				return translate( 'This Post', { comment: 'Filter label for media list' } );
```
Comment (jblz) [inline client/my-sites/media-library/filter-bar.jsx]: On the Pages list, this & other places that mention `Post` should say `Page` instead

Comment (a8ci18n) [inline client/my-sites/media-library/filter-bar.jsx]: Hi! I've found a possible matching string that has already been translated **2** times:
`translate( 'this post' )` *ES Score: 7*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

File: client/my-sites/media-library/list-no-content.jsx [hunk 3]
```diff
@@ -21,6 +21,12 @@ export default React.createClass( {
 
 	getLabel() {
 		switch ( this.props.filter ) {
+			case 'this-post':
+				return this.translate( 'There are no media uploaded to this post.', {
```
Comment (a8ci18n) [inline client/my-sites/media-library/list-no-content.jsx]: :information_source: This string is probably long enough that it doesn't need a **context**. Are you sure you don't want to use a translator comment instead?

File: client/my-sites/media-library/filter-bar.jsx (hunk 2 above)
Comment (jblz) [inline client/my-sites/media-library/filter-bar.jsx]: Actually, I just checked the core behavior and it says `post` regardless of post type (including CPTs), so this is probably fine for the purposes of this PR.

File: client/my-sites/media-library/list-no-content.jsx [hunk 4]
```diff
@@ -10,57 +11,65 @@ import EmptyContent from 'components/empty-content';
...
 	getLabel() {
-		switch ( this.props.filter ) {
+		const {
+			filter,
+			translate,
+		} = this.props;
+
+		switch ( filter ) {
+			case 'this-post':
+				return translate( 'There are no media uploaded to this post.', {
```
Comment (a8ci18n) [inline client/my-sites/media-library/list-no-content.jsx]: :information_source: This string is probably long enough that it doesn't need a **context**. Are you sure you don't want to use a translator comment instead?

File: client/my-sites/media-library/filter-bar.jsx (hunk 2 above)
Comment (jblz) [inline client/my-sites/media-library/filter-bar.jsx]: RE: `this post` vs. `This Post`, capitalizing visually makes sense with the rest of the filtering options, but perhaps not in the localization context.

We can change to the lower-case string here and apply `text-transform: capitalize` in CSS. What do you think, @folletto?

File: client/my-sites/media-library/list-no-content.jsx (hunk 4 above)
Comment (jblz) [inline client/my-sites/media-library/list-no-content.jsx]: Should we change all the `context`s to `comment`s here?

**Comment by drw158** [comment]
Reviewing the comments again. Looks like the segmented control idea was never implemented, but we did agree that it would be better UI/UX.

Mockup:

artboard copy 20

I am willing to just go ahead and merge the current filter solution, just so we can close this gap:

screen shot 2017-07-11 at 3 04 27 pm

Also from rereading the comments it [appears](https://github.com/Automattic/wp-calypso/pull/10466#issuecomment-278157497) that if an image in the post was not uploaded in the post editor, then it will not show up in the "This post" filter. That is very unexpected. I just opened up a post with many images, and none of them showed up in the "this post" filter.

**Comment by gwwar** [comment]
> Depending on how common this is (I'm still not sure why my images were not in the filter) it might be worth putting this feature on hold until we figure it out.

The last time I tested this in wp-admin, it correctly handles this case. It may help to see what we're doing in the wpcom code here.

**Comment by jblz** [comment]
> if an image in the post was not uploaded in the post editor, then it will not show up in the "This post" filter. That is very unexpected. I just opened up a post with many images, and none of them showed up in the "this post" filter

How were the images added to the post, @drw158? Mobile app?

**Comment by jblz** [comment]
For comparison-sake, the wp-admin media picker is calling `/wp-admin/admin-ajax.php` with data that looks like:
`action=query-attachments&post_id=:postid&query%5Borderby%5D=menu_order+ID&query%5Bpost_parent%5D=:postid&query%5Border%5D=ASC&query%5Bposts_per_page%5D=40&query%5Bpaged%5D=1`

The Calypso media picker is calling:
`https://public-api.wordpress.com/rest/v1.1/sites/:siteid/media?http_envelope=1&number=20&post_ID=:postid`

**Comment by jblz** [comment]
> I am willing to just go ahead and merge the current filter solution

Thanks. The UX in your mockup _does_ seem an improvement, but it's also a departure from core behavior:

screen shot 2017-07-11 at 5 14 37 pm

**Comment by drw158** [comment]
I think the images were uploaded with the mobile app.

**Comment by jblz** [comment]
> I think the images were uploaded with the mobile app.

Just confirmed that images uploaded by the Android app are not getting set as a child of the pertinent post -- that seems like a bug to me :-/

Neither the core media picker nor Calypso's are surfacing them with this filter applied.

> I also think that images uploaded via the library don't work either.

This isn't actually a bug as the images are rightfully not children of the pertinent post. It's just...suboptimal as it's not apparent that behind the scenes, the "attachment" is an "orphan."

> That is very unexpected

I agree! This view in the core media picker is confusing at best when including media that weren't uploaded using the editor.

It'd be possible to add a parameter to the `/v1.1/sites/:siteid/media` call to tell it to parse the post's content & include images which are in the site's library but are not a child to the post -- or maybe some enhancement on the core REST API's media endpoint.

> it might be worth putting this feature on hold until we figure it out.

From a "filling the gaps" perspective, this PR seems to check the box. It's pretty much a port of the core experience. I agree, however, there's plenty of room for improvement here -- I'm just not sure when we'd get to implementing such.

**Comment by folletto** [comment]
> Just confirmed that images uploaded by the Android app are not getting set as a child of the pertinent post -- that seems like a bug to me :-/

Yeah, that's the case. I'm not sure where the problem is exactly tho. 
This is the issue: https://github.com/wordpress-mobile/WordPress-Android/issues/5912


> It's pretty much a port of the core experience. I agree, however, there's plenty of room for improvement here -- I'm just not sure when we'd get to implementing such.

I agree with both that it's important to ship this asap, and iterate.
The ideal iteration up is the pill as Dave mocked above. :)

**Comment by drw158** [comment]
UI aside, as much as I want to see this ship and cross it off the list, this isn't something I want our users to use. Sure, it fills a gap we have with WP Admin, but apparently the feature itself is lacking and prone to cause confusion.  If the core experience is not good, we shouldn't port it over just because it's a core feature. We want to make Calypso _better_ than WP Admin.

I don't expect anyone will iterate on this to improve the feature in the near future. Even if the UI I design gets implemented, it seems there are some 'quirks' (as discussed above) that make the feature appear broken to me.

That being said, I know we have issues to close and I'm not that tied to the issue; I just helped with the UI a little bit. I'm ok with what you decide.

**Comment by folletto** [comment]
I hear @drw158 concerns, and I agree we are limited by either mobile bugs as the above, or limitations in Core (which associates an image with a single post, even if it gets used multiple times). These are things we should address separately from this PR.

Back to this PR...

I did a round of testing on Chrome:

screen shot 2017-07-13 at 17 41 55

And seems working well. Also, as much as I'd rather have the pill UI, now there were updates and we have the space available UI there, which cuts out some space, so I think the "This Post" tab, even if not ideal, is something that can be shipped now.

Let's move things forward. For the scope of this PR, the design review is 👍

File: client/my-sites/media-library/filter-bar.jsx (hunk 2 above)
Comment (marekhrabe) [inline client/my-sites/media-library/filter-bar.jsx]: I think capitalize might screw up some translations where they might want a small letter for preposition or something similar. Not exactly sure about this

File: client/my-sites/media-library/list-no-content.jsx (hunk 4 above)
Comment (hewsut) [inline client/my-sites/media-library/list-no-content.jsx]: Yes. Adding context is only recommended when you have a definitive requirement to firmly differentiate between ambiguous variants of a string. Usually the strings will be something short like "Post" (eg, the noun v the verb).  A comment, however, is quite welcome!

Aside: Is it a conscientious choice to treat "media" that way in the copy or does "There isn't any media uploaded to the post." make sense?

File: client/my-sites/media-library/filter-bar.jsx (hunk 2 above)
Comment (hewsut) [inline client/my-sites/media-library/filter-bar.jsx]: My two cents: KISS and don't sweat the new capitalized translation. Re-use--if it fits naturally--is great! But, I wouldn't recommend jumping through hoops over something like this.

**Pull request merged**